fdbserver/VersionedBTree.actor.cpp

```diff
@@ -7667,14 +7667,99 @@ class VersionedBTree {
+		EncryptCipherDomainName domainName;
+		EncryptCipherDomainId domainId = self->getEncryptionDomainId(begin, end, &domainName);
+		TextAndHeaderCipherKeys cipherKeys = wait(getLatestEncryptCipherKeysForDomain(self->db, domainId, domainName));
+		EncryptionKey s;
+		s.cipherKeys = cipherKeys;
+		return s;
+	}
+
+	virtual Future<EncryptionKey> getByRange(const KeyRef& begin, const KeyRef& end) override {
+		return getByRange(this, begin, end);
+	}
+
+	virtual void setTenantPrefixIndex(Reference<TenantPrefixIndex> tenantPrefixIndex) override {
+		this->tenantPrefixIndex = tenantPrefixIndex;
+	}
+
+private:
+	EncryptCipherDomainId getEncryptionDomainId(const KeyRef& begin,
+	                                            const KeyRef& end,
+	                                            EncryptCipherDomainName* domainName) {
+		int64_t domainId = SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID;
+		int64_t beginTenantId = getTenant(begin, true /*inclusive*/);
+		int64_t endTenantId = getTenant(end, false /*inclusive*/);
+		if (beginTenantId == endTenantId && beginTenantId != SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID) {
+			ASSERT(tenantPrefixIndex.isValid());
+			Key tenantPrefix = TenantMapEntry::idToPrefix(beginTenantId);
+			auto view = tenantPrefixIndex->atLatest();
+			auto itr = view.find(tenantPrefix);
+			if (itr != view.end()) {
+				*domainName = *itr;
+				domainId = beginTenantId;
+			} else {
+				// TODO(yiwu): assert this couldn't happen, after we change to only allow encryption when tenant
+				// mode is in required mode.
+				TraceEvent(SevWarn, "EncryptionKeyProvider_TenantNameMissing").detail("TenantID", beginTenantId);
+			}
+		}
+		if (domainId == SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID) {
+			*domainName = FDB_DEFAULT_ENCRYPT_DOMAIN_NAME;
+		}
+		return domainId;
+	}
+
+	int64_t getTenant(const KeyRef& key, bool inclusive) {
+		// A valid tenant id is always a valid encrypt domain id.
+		static_assert(ENCRYPT_INVALID_DOMAIN_ID < 0);
+		if (key.size() < TENANT_PREFIX_SIZE || key >= systemKeys.begin) {
+			return SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID;
+		}
+		// TODO(yiwu): Use TenantMapEntry::prefixToId() instead.
+		int64_t tenantId = bigEndian64(*reinterpret_cast<const int64_t*>(key.begin()));
+		if (tenantId < 0) {
+			return SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID;
+		}
+		if (!inclusive && key.size() == TENANT_PREFIX_SIZE) {
+			tenantId = tenantId - 1;
+		}
+		ASSERT(tenantId >= 0);
+		return tenantId;
+	}
+
+	Reference<AsyncVar<ServerDBInfo> const> db;
+	Reference<TenantPrefixIndex> tenantPrefixIndex;
+};
+
 #include "fdbserver/art_impl.h"
 
 RedwoodRecordRef VersionedBTree::dbBegin(LiteralStringRef(""));
 RedwoodRecordRef VersionedBTree::dbEnd(LiteralStringRef("\xff\xff\xff\xff\xff"));
 
 class KeyValueStoreRedwood : public IKeyValueStore {
 public:
-	KeyValueStoreRedwood(std::string filename, UID logID)
+	KeyValueStoreRedwood(std::string filename, UID logID, Reference<AsyncVar<ServerDBInfo> const> db)
```

Comment:
The point of the `IEncryptionKeyProvider` interface was meant to be that this is all Redwood needs to perform encryption and decryption.  As such, unless it is not possible for some reason the Redwood constructor should just accept `std::shared_ptr<IEncryptionKeyProvider>` and it is up to the creator of the instance to pass an appropriate key provider and set its options.

Reply:
I agree it looks better to pass `EncryptionKeyProvider` into Redwood constructor, but there's an issue. It seems when the worker server opens its local storage engine initially, its `ServerDBInfo` is not updated yet and so we don't know the current tenant mode. And we depend on tenant mode to decide whether to enable encryption or not (which is not done in this PR though). So instead, we will need to rely on the encoding type stored in Redwood to decide whether to enable encryption (which is also not done). So we need to have Redwood to decide which `EncryptionKeyProvider` to use.

An alternative would be to have the worker server pass a map of `EncryptionKeyProvider` for each encoding type and let Redwood choose from them.

Reply:
I don't like either of these options.  Let's talk about this offline.

Reply:
Given we only have one real `EncryptionKeyProvider` except for the null provider, I think we can pass an instance of it to Redwood constructor, and let Redwood decide between it and a null provider (that Redwood create on its own). What do you think?